File: internal/probe/quic/quic.go
package quicprobe

import (
	"context"
	"errors"
	"net"
	"strings"
	"time"

	"icurl/internal/evidence"
	"icurl/internal/probe"
	"icurl/internal/request"
)

type Runner interface {
	Do(context.Context, request.Config) (request.Result, error)
}

type Probe struct {
	Runner Runner
}

func (p Probe) Run(ctx context.Context, target probe.Target) evidence.ProbeResult {
	started := time.Now()
	result := evidence.ProbeResult{
		ProbeName:  "QUICHTTP3",
		Layer:      evidence.LayerQUIC,
		Target:     target.URL.String(),
		StartedAt:  &started,
		Confidence: evidence.ConfidenceObserved,
	}

	if p.Runner == nil {
		finish(&result, evidence.ResultFailed, evidence.ErrorUnknown, "QUIC runner is nil")
		return result
	}

	var response request.Result
	var err error

	ips := target.IPs
	if len(ips) == 0 {
		response, err = p.Runner.Do(ctx, request.Config{
			URL:            target.URL.String(),
			Host:           target.Host, // Pass explicit host for SNI
			Protocol:       request.ProtocolHTTP3Only,
			MaxTime:        10 * time.Second,
			ConnectTimeout: 5 * time.Second,
		})
	} else {
		for _, ip := range ips {
			u := *target.URL
			u.Host = net.JoinHostPort(ip.String(), target.Port)
			response, err = p.Runner.Do(ctx, request.Config{
				URL:            u.String(),
				Host:           target.Host, // Pass explicit host for SNI
				Protocol:       request.ProtocolHTTP3Only,
				MaxTime:        10 * time.Second,
				ConnectTimeout: 5 * time.Second,
			})
			if err == nil {
				break
			}
		}
	}

	if err != nil {
		result.Observations = []string{"HTTP/3 over QUIC did not complete"}
		kind := evidence.ErrorUnknown
		res := evidence.ResultFailed
		errStr := err.Error()
		if errors.Is(err, context.DeadlineExceeded) || strings.Contains(errStr, "timeout") || strings.Contains(errStr, "deadline") {
			kind = evidence.ErrorTimeout
			res = evidence.ResultTimeout
		} else if strings.Contains(errStr, "refused") {
			kind = evidence.ErrorRefused
		} else if strings.Contains(errStr, "reset") {
			kind = evidence.ErrorReset
		} else if strings.Contains(errStr, "certificate") || strings.Contains(errStr, "x509") {
			kind = evidence.ErrorCertificate
		}
		finish(&result, res, kind, err.Error())
		return result
	}

	if response.Body != nil {
		defer func() {
			_ = response.Body.Close()
		}()
	}

	result.Observations = []string{"HTTP/3 request succeeded over " + response.Protocol}
	finish(&result, evidence.ResultOK, evidence.ErrorNone, "")
	return result
}

func finish(result *evidence.ProbeResult, probeResult evidence.Result, kind evidence.ErrorKind, message string) {
	finished := time.Now()
	result.FinishedAt = &finished
	result.DurationNS = finished.Sub(*result.StartedAt).Nanoseconds()
	result.Result = probeResult
	result.ErrorKind = kind
	result.ErrorMessage = message
}

File: icurl-rs/src/probe/quic.rs
use crate::evidence::{classify_error_string, ErrorKind, Layer, ProbeResult, ProbeResultStatus};
use crate::probe::{BoxFuture, Probe, Target};
use crate::request::{Config as ReqConfig, Protocol as ReqProtocol, execute as request_execute};
use std::time::Duration;

pub struct QuicProbe;

impl Probe for QuicProbe {
    fn probe_name(&self) -> &'static str { "QUICHTTP3" }

    fn run<'a>(&'a self, target: &'a Target) -> BoxFuture<'a, ProbeResult> {
        Box::pin(async move {
            let mut result = ProbeResult::new("QUICHTTP3", Layer::Quic, target.url.as_ref());

            let mut response = None;
            let mut last_err = None;

            let ips = &target.ips;
            if ips.is_empty() {
                let req_cfg = ReqConfig {
                    url: target.url.to_string(),
                    host: target.host.clone(),
                    protocol: ReqProtocol::Http3Only,
                    max_time: Duration::from_secs(10),
                    connect_timeout: Duration::from_secs(5),
                    ..Default::default()
                };
                match request_execute(&req_cfg).await {
                    Ok(res) => response = Some(res),
                    Err(e) => last_err = Some(e.to_string()),
                }
            } else {
                for ip in ips {
                    let mut u = target.url.clone();
                    let _ = u.set_host(Some(&ip.to_string()));
                    let _ = u.set_port(Some(target.port));

                    let req_cfg = ReqConfig {
                        url: u.to_string(),
                        host: target.host.clone(),
                        protocol: ReqProtocol::Http3Only,
                        max_time: Duration::from_secs(10),
                        connect_timeout: Duration::from_secs(5),
                        ..Default::default()
                    };
                    match request_execute(&req_cfg).await {
                        Ok(res) => {
                            response = Some(res);
                            break;
                        }
                        Err(e) => {
                            last_err = Some(e.to_string());
                        }
                    }
                }
            }

            if let Some(res) = response {
                result.add_observation(&format!("HTTP/3 request succeeded over {}", res.protocol));
                result.finish(ProbeResultStatus::Ok, ErrorKind::None, "");
            } else {
                result.add_observation("HTTP/3 over QUIC did not complete");
                let err_str = last_err.unwrap_or_else(|| "QUIC/HTTP/3 request failed".to_string());
                let (status, kind) = classify_error_string(&err_str);
                result.finish(status, kind, &err_str);
            }

            result
        })
    }
}
